src/stackmap_checker/trace.c

```diff
@@ -0,0 +1,161 @@
+#include <stdint.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <err.h>
+#include "stmap.h"
+#include "utils.h"
+
+// the label to jump to whenever a guard fails
+extern void restore_and_jmp(void);
+
+// The address to jump to. This represents the address execution should resume
+// at.
+uint64_t addr = 0;
+
+// The saved registers.
+uint64_t r[16];
+
+/*
+ * The guard failure handler.
+ *
+ * This is supposed to restore the stack/register state and resume execution in
+ * the unoptimized version of the trace.
+ */
+void __guard_failure(int64_t sm_id)
+{
+    // Save register the state in global array r
+    asm volatile("mov %%rax,%0\n"
+                 "mov %%rcx,%1\n"
+                 "mov %%rdx,%2\n"
+                 "mov %%rbx,%3\n"
+                 "mov %%rsp,%4\n"
+                 "mov %%rbp,%5\n"
+                 "mov %%rsi,%6\n"
+                 "mov %%rdi,%7\n" : "=m"(r[0]), "=m"(r[1]), "=m"(r[2]),
+                                    "=m"(r[3]), "=m"(r[4]), "=m"(r[5]),
+                                    "=m"(r[6]), "=m"(r[7]) : : );
+    asm volatile("mov %%r8,%0\n"
+                 "mov %%r9,%1\n"
+                 "mov %%r10,%2\n"
+                 "mov %%r11,%3\n"
+                 "mov %%r12,%4\n"
+                 "mov %%r13,%5\n"
+                 "mov %%r14,%6\n"
+                 "mov %%r15,%7\n" : "=m"(r[8]), "=m"(r[9]), "=m"(r[10]),
+                                    "=m"(r[11]), "=m"(r[12]), "=m"(r[13]),
+                                    "=m"(r[14]), "=m"(r[15]) : : );
+    printf("Guard %ld failed!\n", sm_id);
+    void *stack_map_addr = get_addr(".llvm_stackmaps");
+    if (!stack_map_addr) {
+        errx(1, ".llvm_stackmaps section not found. Exiting.\n");
+    }
+
+    stack_map_t *sm = stmap_create(stack_map_addr);
+
+    // The frame address of `trace`.
+    void *bp = __builtin_frame_address(1);
+
+    // Retrieve the indices of the two stack map records which correspond to
+    // the point where a guard failed.
+    // unopt_rec_idx is the index of the stack map record of the unoptimized
+    // version of the function in which a guard failed
+    int unopt_rec_idx = stmap_get_map_record(sm, ~sm_id);
+    // The index of the record which corresponds to the patchpoint which
+    // called __guard_failure.
+    int opt_rec_idx = stmap_get_map_record(sm, sm_id);
+
+    if (unopt_rec_idx == -1 || opt_rec_idx == -1) {
+        errx(1, "Stack map record not found. Exiting.\n");
+    }
+
+    stack_map_record_t unopt_rec = sm->stk_map_records[unopt_rec_idx];
+    stack_map_record_t opt_rec = sm->stk_map_records[opt_rec_idx];
+
+    // Save the direct (stack) locations. These need to be restored later,
+    // because the process of restoring them might cause other values on the
+    // stack of `trace` to be overwritten.
+    uint64_t *direct_locations =
+        (uint64_t *)malloc(sizeof(uint64_t) * unopt_rec.num_locations);
```

Comment:
Do you know about `calloc(3)`?

```
     Designed for safe allocation of arrays, the reallocarray() function is
     similar to realloc() except it operates on nmemb members of size size and
     checks for integer overflow in the calculation nmemb * size.
```

If you were to use `calloc` then you are protected against your multiply overflowing, and potential heap overflow attacks.

Reply:
That is pretty neat, I'll try it out!